[AArch64] Add scheduling information for flag manipulation instructions in Neoverse-V2

PhilippRados opened the pull request:
Fixes #122124

I changed the utilized pipeline to pipeline F (and the flagset) to match the SWOG specification. In order to achieve a throughput=1 even when using multiple num-units I increased the `ReleaseAtCycles` field of both F- and Flag-ProcResources in a new separate def.
I also added these instructions to the Neoverse-V2 lit-tests.

One thing I noticed is that all other defs that set a specific `ReleaseAtCycles` value, set it to be lower than the latency. However in this case the latency should still be 1 while the throughput is also 1, which can only be achieved if `ReleaseAtCycles > Latency`. I'm not sure if this is an issue.
Another thing is that the resource pressure got larger and I'm not sure if this is wanted or if the increase in `ReleaseAtCycles` is too severe.

Since this same format is used in the "Convert floating-point condition flags" instructions (AXFLAG, XAFLAG) it would be easy to add the scheduling info for those instructions too, if needed.

PhilippRados committed:
[AArch64] Added scheduling information for flag manipulation instructions in Neoverse-V2

Rin18 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td:
```diff
@@ -99,6 +99,8 @@ def V2Write_0c : SchedWriteRes<[]> { let Latency = 0; }
 def V2Write_1c_1F    : SchedWriteRes<[V2UnitF]>   { let Latency = 1; }
 def V2Write_1c_1F_1Flg : SchedWriteRes<[V2UnitF, V2UnitFlg]>   { let Latency = 1; }
+def V2Write_1c_1F_1Flg_1T : SchedWriteRes<[V2UnitF, V2UnitFlg]>   { let Latency = 1;
+                                                         let ReleaseAtCycles = [4, 3]; }
```
```suggestion
                                                         let ReleaseAtCycles = [4, 1]; }
```
 I think the resource pressure should be split more evenly for the flags port.

Rin18 replied:
Hi, when I said split more evenly, I meant lower resource pressure.
From the definition of `def V2UnitFlg : ProcResource<3>;  // Flags` , we can see there are 3 resources available for use. Using `let ReleaseAtCycles = [4, 3];` means all 3 resources of `V2UnitFlg` are used for a flag manipulation instruction. Instead, we should lower that pressure, 1 resource per instruction should be enough.

Asher8118 reviewed (changes requested):
Thank you for your patch! It's looking good. I've got a few comments, but nothing major.

>One thing I noticed is that all other defs that set a specific `ReleaseAtCycles` value, set it to be lower than the latency.

The relationship is between the ReleaseAtCycles and its associated Resource, not the latency, so that shouldn't be a problem. 

>Another thing is that the resource pressure got larger and I'm not sure if this is wanted or if the increase in `ReleaseAtCycles` is too severe.

This behaviour is expected to increase resource pressure for pipeline F.
Let's work through what changed. You've added 4 instructions to the `V2-basic-instructions.s` tests, each of which has a latency and throughput of 1. Since the flag setting instructions use pipeline F, which is formed of `V2UnitM0, V2UnitM1, V2UnitS0 and V2UnitS1`, the resource pressure therefore increased by 4 for each of them.

>instructions (AXFLAG, XAFLAG) it would be easy to add the scheduling info for those instructions too, if needed.

Feel free to add these instructions to this PR as well. Make sure to also change the title and commit description if you do add `AXFLAG` and `XAFLAG`.

nit: It'd be preferred to follow commit conventions and keep the commit title in the imperative and use `Add` instead of `Added`. If you haven't already, it'd be good to have a read through the [LLVM Developer policy ](https://llvm.org/docs/DeveloperPolicy.html).

PhilippRados committed:
[AArch64] Lower resource pressure for the flags port

PhilippRados committed:
[AArch64] Add scheduling information for AXFLAG and XAFLAG instructions in Neoverse-V2

PhilippRados commented:
I added the scheduling info for the AXFLAG and XAFLAG instructions. I also changed the instruction matching to not use regex but the instruction definitions directly. From what I've gathered regex is only used when there are multiple very similar instructions that are defined with defm.
One thing I've noticed when trying to remove it was the lingering `def : WriteRes<WriteSys, []> { let Latency = 1; }` on line 1140. This is now mixed in between proper scheduling informations and it might be better to put it somewhere more noticeable. But this is just nit-picking.
Let me know if this is fine like this @Rin18.

Asher8118 reviewed (approved):
LGTM!

Before merging, might be worth checking that the `BOLT.runtime/X86/fdata-escape-chars.ll` test passes after you rebase with `main`.
